mixer/pkg/runtime2/config/ephemeral.go

```diff
@@ -0,0 +1,370 @@
+			continue
+		}
+
+		instanceName := key.String()
+
+		if log.DebugEnabled() {
+			log.Debugf("Processing incoming instance config: name='%s'\n%s", instanceName, resource.Spec.String())
+		}
+
+		cfg := &Instance{
+			Name:     instanceName,
+			Template: info,
+			Params:   resource.Spec,
+		}
+
+		instances[cfg.Name] = cfg
+	}
+
+	counters.instanceConfig.Add(float64(len(instances)))
+	return instances
+}
+
+func (e *Ephemeral) processRuleConfigs(
+	handlers map[string]*Handler,
+	instances map[string]*Instance,
+	counters Counters) []*Rule {
+
+	log.Debug("Begin processing rule configurations.")
+
+	var rules []*Rule
+
+	for ruleKey, resource := range e.entries {
+		if ruleKey.Kind != RulesKind {
+			continue
+		}
+		counters.ruleConfig.Add(1)
+
+		ruleName := ruleKey.String()
+
+		cfg := resource.Spec.(*config.Rule)
+
+		if log.DebugEnabled() {
+			log.Debugf("Processing incoming rule: name='%s'\n%s", ruleName, cfg.String())
+		}
+
+		// TODO(Issue #2139): resourceType is used for backwards compatibility with labels: [istio-protocol: tcp]
+		// Once that issue is resolved, the following block should be removed.
+		rt := resourceType(resource.Metadata.Labels)
+		if cfg.Match != "" {
+			m, err := expr.ExtractEQMatches(cfg.Match)
+			if err != nil {
+				log.Warnf("ConfigWarning: Unable to extract resource type from rule: name='%s'", ruleName)
+
+				// instead of skipping the rule, add it to the list. This ensures that the behavior will
+				// stay the same when this block is removed.
+			} else {
+				if ContextProtocolTCP == m[ContextProtocolAttributeName] {
+					rt.protocol = protocolTCP
+				}
+			}
+		}
+
+		// extract the set of actions from the rule, and the handlers they reference.
+		actions := make([]*Action, 0, len(cfg.Actions))
+		for i, a := range cfg.Actions {
+			log.Debugf("Processing action: %s[%d]", ruleName, i)
+
+			var found bool
+			var handler *Handler
+			handlerName := canonicalize(a.Handler, ruleKey.Namespace)
+			if handler, found = handlers[handlerName]; !found {
+				log.Warnf("handler not found: handler='%s', action='%s[%d]'",
+					handlerName, ruleName, i)
+				counters.ruleConfigError.Inc()
+				continue
+			}
+
+			// Keep track of unique instances, to avoid using the same instance multiple times within the same
+			// action
+			uniqueInstances := make(map[string]bool)
+
+			actionInstances := make([]*Instance, 0, len(a.Instances))
+			for _, instanceName := range a.Instances {
+				instanceName = canonicalize(instanceName, ruleKey.Namespace)
+				if _, found = uniqueInstances[instanceName]; found {
+					log.Warnf("action specified the same instance multiple times: action='%s[%d]', instance='%s',",
+						ruleName, i, instanceName)
+					counters.ruleConfigError.Inc()
+					continue
+				}
+				uniqueInstances[instanceName] = true
+
+				var instance *Instance
+				if instance, found = instances[instanceName]; !found {
+					log.Warnf("ConfigWarning instance not found: instance='%s', action='%s[%d]'",
```

Comment:
message style is not consistent with other log messages.

As above, is the only way we get here because of eventual consistency update issues? If so, we probably shouldn't log a warning here.